Update our team website with current ambition + ADR on systemcard storage (#256)
MinBZK/ai-validation: docs/projects/amt/adrs/0008-systemcard-storage.md

# AMT-0008 System card Storage

## Context

By default, Kubernetes pods use ephemeral storage, which is tied to the pod's lifecycle.
When the pod terminates or restarts, all data is lost. The `/tmp/` directory, being part of
the system's temporary file storage, is cleared during reboots or pod restarts, resulting in
the deletion of system_cards. Therefore, we need a different kind of storage to preserve the data.

## Assumptions

> * The system card data is small to moderate in size and fits within the capabilities of Postgres' JSONB storage.

Reviewer: Optionally consider making this more explicit: what is the average and maximum file size we expect?

## Consequences

### Positive

* **Fast implementation**: The solution is easy to set up, reducing the time to get the project operational.
* **Future proof**: This approach is designed with future scalability in mind. While system cards will initially
be stored in Postgres as JSONB blobs, we anticipate migrating to a Git-based local or remote storage solution
as the system evolves. Importantly, this initial decision allows for a seamless transition in the future,
ensuring no obstacles to migration.
* **Single source & Fast access**: Centralizing everything in a single Postgres database streamlines backups,
reduces maintenance complexity, and ensures quick data access.
* **Built-in permissions**: Postgres provides built-in access control and security through its permission system.

### Negative

* **Data tracking**: Changes to the system card are overwritten, making it difficult to maintain a history or audit trail.
* **Complex queries**: Complex queries can be inefficient and require custom parsing.
* **Collaboration**: Collaborating on the JSONB data is challenging due to its complex format and lack of version control.

> * **Scalability**: As the JSONB blobs grow in size, the storage overhead and query performance may become significant issues.

Reviewer: Do we already know if BJSON blob storage is supported by sqlite? IT would also be a negative if this is not supported, because it would make local testing more difficult.

# AMT-0008 System card Storage

## Context

By default, Kubernetes pods use ephemeral storage, which is tied to the pod's lifecycle.
When the pod terminates or restarts, all data is lost. The `/tmp/` directory, being part of
the system's temporary file storage, is cleared during reboots or pod restarts, resulting in
the deletion of system_cards. Therefore, we need a different kind of storage to preserve the data.

## Assumptions

> * The system card data is small to moderate in size (up to 2GB), making it manageable for SQLite.

Reviewer: I think I saw somewhere online that Postgres supports binary json up to 255MB. That would not be sufficient if a system card has data up to 2GB. Although 2GB sounds rather large for a system card. (https://www.crunchydata.com/blog/state-of-json-in-postgres-2022)

Reviewer 2: Why do we think it would be in that range? I'd expect system cards to be sub 1MB.

> * The system card data is small to moderate in size (up to 255MB), making it manageable for SQLite.

Reviewer 2: ```suggestion
* The system card data is small to moderate in size (up to 255MB), making it manageable to store in databases (in postgres as well as in  in SQLite).
```

* Tracking changes to the system card data over time is not a priority in the short term, but may become
necessary in the future.

## Decision

> The system card of an algorithm system is stored solely as a JSONB blob in the projects table in Postgres,

Reviewer 2: ```suggestion
The system card of an algorithm system is stored solely as a JSON blob in the projects table in Postgres,
```

with no additional storage elsewhere.

## Risks

* **Data Overwrite**: As the system card is overwritten with each update, it becomes difficult to track
historical changes or revert to previous states.

> * **Scaling**: As the project grows, managing larger JSONB blobs may present performance challenges,

Reviewer 2: ```suggestion
* **Scaling**: As the project grows, managing larger JSON blobs may present performance challenges,
```

particularly when handling complex queries.

> * **Collaboration**: Collaborating on the system card content is more difficult, as the JSONB format

Reviewer 2: ```suggestion
* **Collaboration**: Collaborating on the system card content is more difficult, as the JSON format
```

requires parsing and manual intervention for certain tasks.

> * **Limited Querying**: While Postgres supports querying JSONB, complex queries and data manipulations

Reviewer 2: ```suggestion
* **Limited Querying**: While Postgres supports querying and indexing JSON fields, complex queries and data manipulations
```

be stored in Postgres as JSONB blobs, we anticipate migrating to a Git-based local or remote storage solution
as the system evolves. Importantly, this initial decision allows for a seamless transition in the future,
ensuring no obstacles to migration.
* **Single source & Fast access**: Centralizing everything in a single Postgres database streamlines backups,
reduces maintenance complexity, and ensures quick data access.
* **Built-in permissions**: Postgres provides built-in access control and security through its permission system.

### Negative

* **Data tracking**: Changes to the system card are overwritten, making it difficult to maintain a history or audit trail.
* **Complex queries**: Complex queries can be inefficient and require custom parsing.
* **Collaboration**: Collaborating on the JSONB data is challenging due to its complex format and lack of version control.
* **Scalability**: As the JSONB blobs grow in size, the storage overhead and query performance may become significant issues.
* **Not supported by SQLite**: While SQLite supports JSON through its JSON1 extension, it does not support PostgreSQL's

> JSONB data type natively, which complicates local development and testing environments that rely on SQLite as a database

Reviewer 2: Right, this is why I actually already went for JSON, not JSONB
